numpy/core/tests/test_indexing.py

```diff
@@ -184,19 +184,24 @@ def test_single_bool_index(self):
                       [4, 5, 6],
                       [7, 8, 9]])
 
-        # Python boolean converts to integer
-        # These are being deprecated (and test in test_deprecations)
-        #assert_equal(a[True], a[1])
-        #assert_equal(a[False], a[0])
-
-        # Same with NumPy boolean scalar
-        # Before DEPRECATE, this is an error (as always, but telling about
-        # future change):
-        assert_raises(IndexError, a.__getitem__, np.array(True))
-        assert_raises(IndexError, a.__getitem__, np.array(False))
-        # After DEPRECATE, this behaviour can be enabled:
-        #assert_equal(a[np.array(True)], a[None])
-        #assert_equal(a[np.array(False), a[None][0:0]])
+        assert_equal(a[np.array(True)], a[None])
+        # @seberg, should this work?
+        # assert_equal(a[np.array(False), a[None][0:0]])
```

Comment:
OK, this seems to do the trick:
```diff
@@ -441,16 +441,6 @@ prepare_index(PyArrayObject *self, PyObject *index,
 
             if (PyArray_NDIM(arr) == 0) {
                 /*
-                 * TODO, WARNING: This code block cannot be used due to
-                 *                FutureWarnings at this time. So instead
-                 *                just raise an IndexError.
-                 */
-                PyErr_SetString(PyExc_IndexError,
-                        "in the future, 0-d boolean arrays will be "
-                        "interpreted as a valid boolean index");
-                Py_DECREF((PyObject *)arr);
-                goto failed_building_indices;
-                /*
                  * This can actually be well defined. A new axis is added,
                  * but at the same time no axis is "used". So if we have True,
                  * we add a new axis (a bit like with np.newaxis). If it is
@@ -459,7 +449,6 @@ prepare_index(PyArrayObject *self, PyObject *index,
 
                 index_type |= HAS_FANCY;
                 indices[curr_idx].type = HAS_0D_BOOL;
-                indices[curr_idx].value = 1;
 
                 /* TODO: This can't fail, right? Is there a faster way? */
                 if (PyObject_IsTrue((PyObject *)arr)) {
@@ -468,6 +457,7 @@ prepare_index(PyArrayObject *self, PyObject *index,
                 else {
                     n = 0;
                 }
+                indices[curr_idx].value = n;
                 indices[curr_idx].object = PyArray_Zeros(1, &n,
                                             PyArray_DescrFromType(NPY_INTP), 0);
                 Py_DECREF(arr);
@@ -2450,6 +2440,10 @@ mapiter_fill_info(PyArrayMapIterObject *mit, npy_index_info *indices,
             mit->fancy_dims[j] = 1;
             /* Does not exist */
             mit->iteraxes[j++] = -1;
+            if (mit->dimensions[mit->nd_fancy-1] > 1) {
+                goto broadcast_error;
+            }
+            mit->dimensions[mit->nd_fancy-1] *= indices[i].value;
         }
 
         /* advance curr_dim for non-fancy indices */
@@ -2487,7 +2481,7 @@ mapiter_fill_info(PyArrayMapIterObject *mit, npy_index_info *indices,
     }
 
     for (i = 0; i < index_num; i++) {
-        if (indices[i].type != HAS_FANCY) {
+        if (!(indices[i].type & HAS_FANCY)) {
             continue;
         }
         tmp = convert_shape_to_string(
```

I still suggest to split it off, it likely warrents more new tests, and another look.